### PR TITLE
[WIP] Rewrite roads layers

### LINKED ISSUE #2046: Combine road lines and areas to one layer
From https://github.com/gravitystorm/openstreetmap-carto/issues/1975#issuecomment-160467512

Doing this will reduce the number of layers and allow future MSS simplifications

### LINKED ISSUE #2468: roads.mss carto code style error
At line 1342

```mss
    [feature = 'highway_residential'],
    [feature = 'highway_unclassified'] {
      [zoom = 12][feature = 'highway_residential'] {
        line-color: @unimportant-road;
        line-width: 0.4;
      }
      [zoom = 12][feature = 'highway_unclassified'] {
        line-color: @unimportant-road;
        line-width: 1;
      }
```

`zoom =` should [never be used](https://github.com/gravitystorm/openstreetmap-carto/blob/master/CONTRIBUTING.md#cartocss-style-guidelines)

### PR DESCRIPTION
**Pull request by pnorman**
Fixes #2046
Fixes #2468

This is a complete rewrite of the transportation code. The primary goals are

- Go from 9 road layers to 2
- Get all the MSS for each feature in the same place
- Avoid selecting from `planet_osm_line` once for roads and once for railways
- Reduce MSS duplication

Once I'm farther along I'll set up a demo, but there are still features not yet implemented.

The layer changes are to move barrier-like features, landuse overlays, and `ferry-routes` before roads, and to move `aerialways` and `waterway-bridges` after roads.

## Cartographic changes

- Non-motorized paths (footway/path, cycleway, and bridleway) use the footway width
- Various minor adjustments of start zooms for casings, bridges, etc to be consistent.

## Remaining work

- [x] More railways
- [ ] Tunnels
- [ ] Oneway arrows
- [ ] Access
- [ ] Low zoom roads
- [ ] End caps
- [ ] Unify symbolizer names

Long-term I'd like to redo names, but intend to keep that out of this PR

## Known issues

- [ ] Footway weight has changed.

I'm not sure what has caused this, but unknown surface footways are lighter.

image

For testing, areas likely to have problems will involve

- Endcaps
- Complex layering

### CODE REVIEW

File: transportation.mss [hunk 1]
```diff
@@ -0,0 +1,1092 @@
...
+          marker-width: @residential-width-z19 * 1.6;
+          marker-height: @residential-width-z19 * 1.6;
+        }
+        marker-allow-overlap: true;
+        marker-ignore-placement: true;
+        marker-line-width: 0;
+      }
+    }
+  }
+  [highway = 'pedestrian'],
```
Comment (sommerluk) [inline transportation.mss]: Missing newline above this line

File: project.mml [hunk 2]
```diff
@@ -957,95 +551,152 @@ Layer:
...
+              points.highway AS turning_type,
+              COALESCE(layer,0) AS layernotnull,
+              0 AS z_order -- TODO: handle this somehow?
+            FROM planet_osm_point points
+              JOIN lines ON ST_DWithin(points.way, lines.way, 0.1)
+            WHERE points.way && !bbox!
+              AND points.highway IN ('turning_circle', 'turning_loop')
+              AND lines.highway IN ('tertiary', 'unclassified', 'residential', 'living_street', 'service', 'track')
+              ORDER BY points.osm_id, -- for DISTINCT ON
+                CASE lines.highway -- Use the most important type for cases where the TC is at an intersection. TODO: Should lines.layer be in here?
```
Comment (sommerluk) [inline project.mml]: Is it possible to simply use z_order instead of this multiline-CASE?

File: project.mml [hunk 3]
```diff
@@ -957,95 +551,152 @@ Layer:
...
+                  END
+                ELSE highway END AS highway,
+              railway,
+              construction,
+              substr(highway, length(highway)-4, 5) = '_link' AS link,
+              COALESCE(tunnel IN ('yes', 'building_passage') OR covered = 'yes', FALSE) AS tunnel,
+              COALESCE(bridge IN ('yes', 'boardwalk', 'cantilever', 'covered', 'low_water_crossing', 'movable', 'trestle', 'viaduct'), FALSE) AS bridge,
+              COALESCE(layer,0) AS layernotnull,
+              z_order,
+              surface IN ('paved', 'asphalt', 'cobblestone', 'cobblestone:flattened', 'sett', 'concrete', 'concrete:lanes', 'concrete:plates', 'paving_stones', 'metal', 'wood')
```
Comment (sommerluk) [inline project.mml]: Doesn’t this query always return TRUE for everything that is not in the “unpaved” list anyway, making the query for th e“paved” list obsolete?

Comment (pnorman) [reply project.mml]: No, but there is a bug with values not in either list.

**Comment by pnorman** [comment]
Noting a minor bug

Pedestrian features are rendering above turning circles fill

image

**Comment by sommerluk** [comment]
Maps.me prend plusieurs mois pour actualiser la carte. En Afrique, en général c’est encore plus lent que en Europe.

**Comment by sommerluk** [comment]
Currently, first comes the code for more important features with high z_order (e.g. highway=motorway) and later the code for less important features with low z_order (e.g. highway=track).

Might it be an option to use the opposite order?

This would not make any functional change.

For the code proposed in #2640 (unpaved roads with dst-out pattern), the opposite order would make it a lot easier. In fact, the diff of #2640 is difficult to read because it needs the opposite order because it uses attachments. The code changes themself are not so heavy.

And maybe this is also useful independent from #2640. Indeed the old road code has high z_order features first. But the attachment concept in MSS works the other way around: First attachment drawn first, last attachment drawn last (above the other ones). As we use attachments anyway in our roads MSS code, it could be more intuitive to have the rest following the same logic…

**Comment by pnorman** [comment]
> Currently, first comes the code for more important features with high z_order (e.g. highway=motorway) and later the code for less important features with low z_order (e.g. highway=track).
> 
> Might it be an option to use the opposite order?

I find it easier to read from most important to least important.

> For the code proposed in #2640 (unpaved roads with dst-out pattern), the opposite order would make it a lot easier. In fact, the diff of #2640 is difficult to read because it needs the opposite order because it uses attachments. The code changes themself are not so heavy.

I'd rather keep it as-is for now, then we can look at changing it if necessary.

> As we use attachments anyway in our roads MSS code, it could be more intuitive to have the rest following the same logic…

Generally we order in descending significance for MSS.

It's interesting to note that within the SQL, the order by statements are both ascending and descending, depending on if its drawing lines and fills, or placing labels.

**Comment by sommerluk** [comment]
> It's interesting to note that within the SQL, the order by statements are both ascending and descending, depending on if its drawing lines and fills, or placing labels.

Good point.

**Comment by sommerluk** [comment]
Will the road layer rewrite resolve at the same time #2595 (construction roads should not be wider than their finished counterparts)?

**Comment by pnorman** [comment]
> Will the road layer rewrite resolve at the same time #2595 (construction roads should not be wider than their finished counterparts)?

I'm not intending to make cartographic changes, but that one will depend on how the construction MSS is structured.

**Comment by kocio-pl** [comment]
@pnorman What is the current state of this code? Do you still plan to update it to be mergable (maybe just some part of it, if the task is too big)?

**Comment by pnorman** [comment]
> @pnorman What is the current state of this code?

I've been lacking time and interest.

> Do you still plan to update it to be mergable (maybe just some part of it, if the task is too big)?

I don't really see a way to split it down in scope.

The best way to help this out would be for someone to pair up with me and we could work on it together. This would also be a good chance for someone to learn more about complex CartoCSS. The new roads code is simpler, but roads are still one of our biggest areas.

**Comment by kocio-pl** [comment]
Thanks for the information. Is there anybody wanting to help Paul with this code?

**Comment by matthijsmelissen** [comment]
I'm closing this as currently nobody is interested in working on this.

If anybody is interested, feel free to create a new proposal.

@pnorman Thanks a lot for the work you put in so far! Hopefully somebody else can pick up.

**Pull request closed without merging**